dot/state/grandpa.go

```diff
@@ -24,44 +26,570 @@ var (
+		scheduledChangeHash := scheduledChange.change.announcingHeader.Hash()
+
+		isDescendant, err := s.blockState.IsDescendantOf(finalizedHash, scheduledChangeHash)
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			onBranchScheduledChanges = append(onBranchScheduledChanges, scheduledChange)
+		}
+	}
+
+	s.scheduledChangeRoots = make([]*pendingChangeNode, len(onBranchScheduledChanges))
+	copy(s.scheduledChangeRoots, onBranchScheduledChanges)
+	return nil
+}
+
+// ApplyScheduledChanges will check the schedules changes in order to find a root
+// that is equals or behind the finalized number and will apply its authority set changes
+func (s *GrandpaState) ApplyScheduledChanges(finalizedHeader *types.Header) error {
+	finalizedHash := finalizedHeader.Hash()
+
+	err := s.keepDescendantForcedChanges(finalizedHash, finalizedHeader.Number)
+	if err != nil {
+		return fmt.Errorf("cannot keep descendant forced changes: %w", err)
+	}
+
+	if len(s.scheduledChangeRoots) == 0 {
+		return nil
+	}
+
+	changeToApply, err := s.getApplicableScheduledChange(finalizedHash, finalizedHeader.Number)
+	if err != nil {
+		return fmt.Errorf("cannot finalize scheduled change: %w", err)
+	}
+
+	if changeToApply == nil {
+		return nil
+	}
+
+	logger.Debugf("applying scheduled change: %s", changeToApply)
+
+	newSetID, err := s.IncrementSetID()
+	if err != nil {
+		return fmt.Errorf("cannot increment set id: %w", err)
+	}
+
+	grandpaVotersAuthorities := types.NewGrandpaVotersFromAuthorities(changeToApply.nextAuthorities)
+	err = s.setAuthorities(newSetID, grandpaVotersAuthorities)
+	if err != nil {
+		return fmt.Errorf("cannot set authorities: %w", err)
+	}
+
+	err = s.setChangeSetIDAtBlock(newSetID, changeToApply.effectiveNumber())
+	if err != nil {
+		return fmt.Errorf("cannot set change set id at block")
+	}
+
+	logger.Debugf("Applying authority set change scheduled at block #%d",
+		changeToApply.announcingHeader.Number)
+
+	// TODO: add afg.applying_scheduled_authority_set_change telemetry info here
+	return nil
+}
+
+// ApplyForcedChanges will check for if there is a scheduled forced change relative to the
+// imported block and then apply it otherwise nothing happens
+func (s *GrandpaState) ApplyForcedChanges(importedBlockHeader *types.Header) error {
+	importedBlockHash := importedBlockHeader.Hash()
+	var forcedChange *pendingChange
+
+	for _, forced := range s.forcedChanges {
+		announcingHash := forced.announcingHeader.Hash()
+		effectiveNumber := forced.effectiveNumber()
+
+		if importedBlockHash.Equal(announcingHash) && effectiveNumber == 0 {
+			forcedChange = forced
+			break
+		}
+
+		isDescendant, err := s.blockState.IsDescendantOf(announcingHash, importedBlockHash)
+		if err != nil {
+			return fmt.Errorf("cannot check ancestry: %w", err)
+		}
+
+		if !isDescendant {
+			continue
+		}
+
+		if effectiveNumber == importedBlockHeader.Number {
+			forcedChange = forced
+			break
+		}
```

Comment:
I think no, the consensus digest is a SCALE encoded enum, so the same block cannot have more than one forced change but it can have a finalized and a scheduled change

Reply:
So should we not break and collect all the forced changes?